File: SOLID/LSP(LiskovSubstitutionPrinciple)/LSP_violated.cpp
#include <bits/stdc++.h>
using namespace std;

class Account
{
public:
    virtual void deposit(int amount) = 0;
    virtual void withdraw(int amount) = 0;
};

class SavingAccount : public Account
{
private:
    int balance;

public:
    SavingAccount()
    {
        balance = 0;
    }
    void deposit(int amount)
    {
        balance += amount;
        cout << "Deposited: " << amount << " in the saving Account, New Balance: " << balance << endl;
    }

    void withdraw(int amount)
    {
        if (balance >= amount)
        {
            balance -= amount;
            cout << "Withdrawn: " << amount << " from the saving Accouint. new Balance: " << balance << endl;
        }
        else
        {
            cout << "not enough balance in saving account\n";
        }
    }
};
class CurrentAccount : public Account
{
private:
    int balance;

public:
    CurrentAccount()
    {
        balance = 0;
    }
    void deposit(int amount)
    {
        balance += amount;
        cout << "Deposited: " << amount << " in the current Account, New Balance: " << balance << endl;
    }

    void withdraw(int amount)
    {
        if (balance >= amount)
        {
            balance -= amount;
            cout << "Withdrawn: " << amount << " from the current Accouint. new Balance: " << balance << endl;
        }
        else
        {
            cout << "not enough balance in current account\n";
        }
    }
};
class FixedAccount : public Account
{
private:
    int balance;

public:
    FixedAccount()
    {
        balance = 0;
    }
    void deposit(int amount)
    {
        balance += amount;
        cout << "Deposited: " << amount << " in the fixed Account, New Balance: " << balance << endl;
    }

    void withdraw(int amount)
    {
        throw logic_error("Withdrawl is not allowed!");
    }
};

class BankClient
{
private:
    vector<Account *> accounts;

public:
    BankClient(vector<Account *> accounts) : accounts(accounts) {}

    void processTransactions()
    {
        for (Account *acc : accounts)
        {
            acc->deposit(1000);
            // assuming all accounts support withdrawl

            try
            {
                acc->withdraw(500);
            }
            catch (const logic_error &e)
            {
                cout << "Exception: " << e.what() << endl;
            }
        }
    }
};

int main()
{
    vector<Account *> accounts;
    accounts.push_back(new SavingAccount());
    accounts.push_back(new CurrentAccount());
    accounts.push_back(new FixedAccount());

    BankClient *client = new BankClient(accounts);
    client->processTransactions();
    return 0;
}

File: Coder Army/SOLID/LSP(LiskovSubstitutionPrinciple)/Javascript code/LSP_violated.js
class Account{
  deposit(amount){}
  withdraw(amount){}
}

class SavingAccount extends Account{
  constructor(){
    super()
    this.balance=0
  }
  deposit(amount) {
    this.balance += amount;
    console.log(`Deposited: ${amount} in the saving account, New Balance: ${this.balance}`);
  }

  withdraw(amount) {
    if (this.balance >= amount) {
      this.balance -= amount;
      console.log(`Withdrawn: ${amount} from the saving account. New Balance: ${this.balance}`);
    } else {
      console.log("Not enough balance in saving account");
    }
  }
}

class CurrentAccount extends Account{
  constructor(){
    super()
    this.balance=0
  }
  deposit(amount) {
    this.balance += amount;
    console.log(`Deposited: ${amount} in the current account, New Balance: ${this.balance}`);
  }

  withdraw(amount) {
    if (this.balance >= amount) {
      this.balance -= amount;
      console.log(`Withdrawn: ${amount} from the current account. New Balance: ${this.balance}`);
    } else {
      console.log("Not enough balance in current account");
    }
  }
}

class FixedAccount extends Account{
  constructor(){
    super()
    this.balance=0
  }
  deposit(amount) {
    this.balance += amount;
    console.log(`Deposited: ${amount} in the fixed account, New Balance: ${this.balance}`);
  }

  withdraw(amount) {
    throw new Error("Withdrawal is not allowed!");
  }
}

class BankClient{
  constructor(accounts){
    this.accounts=accounts
  }

  processTransaction(){
    for(const acc of this.accounts){
      acc.deposit(100);
      try {
        acc.withdraw(40)
      } catch (error) {
        console.log(`Exception: ${error.message}`);
        
      }
    }
  }
}

const accounts=[
  new SavingAccount(),
  new CurrentAccount(),
  new FixedAccount
]

const client = new BankClient(accounts)
client.processTransaction()